File: elixir/lib/2017/01_inverse_captcha/aoc201701.ex
defmodule AOC2017.Day01 do
  @moduledoc """
  Advent of Code 2017, day 1: Inverse Captcha
  """
  require AOC
  require Integer

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    puzzle_input
    |> String.to_integer()
    |> Integer.digits()
  end

  @doc """
  Solve part 1
  """
  def part1(input) do
    input
    |> prepend_last()
    |> Enum.reduce({nil, 0}, &compare_consecutive/2)
    |> elem(1)
  end

  @doc """
  Solve part 2
  """
  def part2(input) do
    half_length = input |> length() |> div(2) |> trunc()

    input
    |> Enum.chunk_every(half_length)
    |> Enum.zip()
    |> Enum.filter(fn {first, second} -> first == second end)
    |> Enum.map(fn {first, second} -> first + second end)
    |> Enum.sum()
  end

  @doc """
  Add last digit in front of list

  ## Example:

      iex> prepend_last([1, 2, 3])
      [3, 1, 2, 3]
  """
  def prepend_last(digits), do: [digits |> Enum.reverse() |> hd | digits]

  @doc """
  Add number to running total if it's equal to the previous number

  ## Examples:

      iex> compare_consecutive(3, {3, 42})
      {3, 45}

      iex> compare_consecutive(3, {4, 42})
      {3, 42}
  """
  def compare_consecutive(current, {current, total}), do: {current, total + current}
  def compare_consecutive(current, {_previous, total}), do: {current, total}

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/2017/01_inverse_captcha/aoc201701.py
"""AoC 1, 2017: Inverse Captcha"""

# Standard library imports
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input"""
    return [int(digit) for digit in puzzle_input]


def part1(data):
    """Solve part 1"""
    return equal_sum(data, data[1:] + data[:1])


def part2(data):
    """Solve part 2"""
    split_idx = len(data) // 2
    return equal_sum(data, data[split_idx:] + data[:split_idx])


def equal_sum(sequence_1, sequence_2):
    """Find sum of digits that are equal in both sequences

    >>> equal_sum([1, 2, 3], [1, 2, 3])
    6

    >>> equal_sum([1, 2, 3, 4], [4, 3, 2, 1])
    0
    """
    return sum(
        first for first, second in zip(sequence_1, sequence_2) if first == second
    )


def solve(puzzle_input):
    """Solve the puzzle for the given input"""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
